src/lib/pubkey/classic_mceliece/cmce.cpp

```diff
@@ -0,0 +1,172 @@
+                                                       std::span<const uint8_t> key_bits) :
+      Classic_McEliece_PublicKey(key_bits, cmce_param_set_from_oid(alg_id.oid())) {}
+
+Classic_McEliece_PublicKey::Classic_McEliece_PublicKey(std::span<const uint8_t> key_bits,
+                                                       Classic_McEliece_Parameter_Set param_set) {
+   auto params = Classic_McEliece_Parameters::create(param_set);
+   BOTAN_ARG_CHECK(key_bits.size() == params.pk_size_bytes(), "Wrong public key length");
+   std::vector<uint8_t> key_bits_vec(key_bits.begin(), key_bits.end());
+   m_public = std::make_shared<Classic_McEliece_PublicKeyInternal>(
+      params, Classic_McEliece_Matrix(params, std::move(key_bits_vec)));
+}
+
+Classic_McEliece_PublicKey::Classic_McEliece_PublicKey(const Classic_McEliece_PublicKey& other) {
+   m_public = std::make_shared<Classic_McEliece_PublicKeyInternal>(*other.m_public);
+}
+
+Classic_McEliece_PublicKey& Classic_McEliece_PublicKey::operator=(const Classic_McEliece_PublicKey& other) {
+   if(this != &other) {
+      m_public = std::make_shared<Classic_McEliece_PublicKeyInternal>(*other.m_public);
+   }
+   return *this;
+}
+
+AlgorithmIdentifier Classic_McEliece_PublicKey::algorithm_identifier() const {
+   return AlgorithmIdentifier(object_identifier(), AlgorithmIdentifier::USE_EMPTY_PARAM);
+}
+
+OID Classic_McEliece_PublicKey::object_identifier() const {
+   return m_public->params().object_identifier();
+}
+
+size_t Classic_McEliece_PublicKey::key_length() const {
+   return m_public->matrix().bytes().size();
+}
+
+size_t Classic_McEliece_PublicKey::estimated_strength() const {
+   return m_public->params().estimated_strength();
+}
+
+std::vector<uint8_t> Classic_McEliece_PublicKey::public_key_bits() const {
+   return m_public->matrix().bytes();
+}
+
+bool Classic_McEliece_PublicKey::check_key(RandomNumberGenerator&, bool) const {
+   return true;
+}
+
+std::unique_ptr<Private_Key> Classic_McEliece_PublicKey::generate_another(RandomNumberGenerator& rng) const {
+   return std::make_unique<Classic_McEliece_PrivateKey>(rng, m_public->params().set());
+}
+
+std::unique_ptr<PK_Ops::KEM_Encryption> Classic_McEliece_PublicKey::create_kem_encryption_op(
+   std::string_view params, std::string_view provider) const {
+   if(provider.empty() || provider == "base") {
+      return std::make_unique<Classic_McEliece_Encryptor>(this->m_public, params);
+   }
+   throw Provider_Not_Found(algo_name(), provider);
+}
+
+Classic_McEliece_PrivateKey::Classic_McEliece_PrivateKey(RandomNumberGenerator& rng,
+                                                         Classic_McEliece_Parameter_Set param_set) {
+   auto params = Classic_McEliece_Parameters::create(param_set);
+   auto seed = rng.random_vec<Initial_Seed>(params.seed_len());
+   auto key_pair = Classic_McEliece_KeyPair_Internal::generate(params, seed);
+
+   m_private = key_pair.private_key;
+   m_public = key_pair.public_key;
```

Comment:
We decided against a freestanding function because we liked the explicit initialization of the pk and sk and didn't want to specify a namespace for one function only. As a compromise, I add a function `decompose_to_pair() &&` to the struct to make use of std::tie (see c1dd5ab).

Reply:
Certainly good enough. Thanks.